rust/src/sdk/api.rs

```diff
@@ -174,235 +246,393 @@ pub unsafe extern "C" fn drop_client(client: *mut Client) {
-                client.cached_model = None;
-                [<PrimitiveModel $prim_rust:upper>] { ptr, len }
+                4 => {
+                    if let Some(CachedModel::I64(ref mut cached_model)) = client.cached_model {
+                        return PrimitiveModel {
+                            ptr: cached_model.as_mut_ptr() as *mut c_void,
+                            len: cached_model.len() as c_ulong,
+                            dtype,
+                        };
+                    }
+                }
+                _ => unreachable!(),
             }
         }
-    };
-}
 
-get_model!(f32, c_float, "[`PrimitiveModelF32`]");
-get_model!(f64, c_double, "[`PrimitiveModelF64`]");
-get_model!(i32, c_int, "[`PrimitiveModelI32`]");
-get_model!(i64, c_long, "[`PrimitiveModelI64`]");
-
-/// Generates a method to register the updated local model. The argument `$prim` is the
-/// corresponding Rust primitive data type and `$doc` is a type link for the documentation.
-macro_rules! update_model {
-    ($prim:ty, $doc:expr $(,)?) => {
-        paste::item! {
-            #[allow(unused_unsafe)]
-            #[no_mangle]
-            #[doc = "Registers the updated local model.\n"]
-            #[doc = "\n"]
-            #[doc = "A `model` which doesn't point to memory allocated by"]
-            #[doc = $doc]
-            #[doc = "requires additional copying while beeing iterated over.\n"]
-            #[doc = "\n"]
-            #[doc = "# Safety\n"]
-            #[doc = "The method dereferences from the raw pointer arguments. Therefore, the"]
-            #[doc = "behavior of the method is undefined if the arguments don't point to valid"]
-            #[doc = "objects.\n"]
-            #[doc = "\n"]
-            #[doc = "The `model` points to memory which is either allocated by"]
-            #[doc = $doc]
-            #[doc = "and then modified or which isn't allocated by"]
-            #[doc = $doc]
-            #[doc = ". Therefore, the behavior of the method is undefined if any of the"]
-            #[doc = "[slice safety conditions](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html#safety)"]
-            #[doc = " are violated for `model`."]
-            pub unsafe extern "C" fn [<update_model_ $prim>](
-                client: *mut Client,
-                model: [<PrimitiveModel $prim:upper>],
-            ) {
-                if client.is_null()
-                    || model.ptr.is_null()
-                    || model.len == 0
-                    || model.len > (isize::MAX as usize / mem::size_of::<$prim>()) as c_ulong
-                {
-                    // TODO: add error handling
-                    panic!("invalid primitive model");
+        // deserialize and convert the global model to a primitive model and cache it
+        client.has_new_global_model_since_last_cache = false;
+        if let Ok(deserialized_model) = bincode::deserialize::<Model>(&global_model) {
```

Comment:
Unrelated to this PR of course, but this back and forth between model representations is very unfortunate... We need to address that properly at some point.

Reply:
i agree, it's a lot of allocation. nevertheless, at some point the user needs to have a model of primitive data type to actually train on it. but at least the serde part should be properly splitted from that.